pallets/creditcoin/src/ocw/collect_coins.rs

```diff
@@ -0,0 +1,757 @@
+
+///exchange has been deprecated, use burn instead
+fn burn_vested_cc_abi() -> Function {
+	#[allow(deprecated)]
+	Function {
+		name: "exchange".into(),
+		inputs: vec![Param {
+			name: "value".into(),
+			kind: ParamType::Uint(256),
+			internal_type: None,
+		}],
+		outputs: vec![Param { name: "success".into(), kind: ParamType::Bool, internal_type: None }],
+		constant: false,
+		state_mutability: StateMutability::NonPayable,
+	}
+}
+pub fn validate_collect_coins(
+	to: &ExternalAddress,
+	amount: &mut Option<ExternalAmount>,
+	receipt: &EthTransactionReceipt,
+	transaction: &EthTransaction,
+	eth_tip: U64,
+) -> OffchainResult<()> {
+	ensure!(receipt.is_success(), VerificationFailureCause::TaskFailed);
+	let vesting_contract = hex::decode(b"a3EE21C306A700E682AbCdfe9BaA6A08F3820419").unwrap();
+	let vesting_contract = &H160::from(<[u8; 20]>::try_from(vesting_contract.as_slice()).unwrap());
+
+	let block_number = transaction.block_number.ok_or(VerificationFailureCause::TaskPending)?;
+
+	let diff = (eth_tip).checked_sub(block_number);
+	ensure!(diff.is_some(), VerificationFailureCause::TaskInFuture);
+
+	ensure!(diff.unwrap().as_u64() >= ETH_CONFIRMATIONS, VerificationFailureCause::TaskUnconfirmed);
+
+	if let Some(to) = &transaction.to {
+		ensure!(to == vesting_contract, VerificationFailureCause::IncorrectContract);
+	} else {
+		return Err(VerificationFailureCause::MissingReceiver.into());
+	}
+
+	if let Some(from) = &transaction.from {
+		ensure!(from[..] == to[..], VerificationFailureCause::IncorrectSender)
+	} else {
+		return Err(VerificationFailureCause::MissingSender.into());
+	}
+
+	let transfer_fn = burn_vested_cc_abi();
+	//is ignoring the selector a good idea? test? Same input, diff call (not exchange)?
+	ensure!(transaction.input.0.len() > 4, VerificationFailureCause::EmptyInput);
+	let inputs = transfer_fn.decode_input(&transaction.input.0[4..]).map_err(|e| {
+		log::error!("failed to decode inputs: {:?}", e);
+		VerificationFailureCause::AbiMismatch
+	})?;
+
+	let input_amount = match inputs.get(0) {
+		Some(Token::Uint(value)) => ExternalAmount::from(value),
+		_ => return Err(VerificationFailureCause::IncorrectInputType.into()),
+	};
+
+	*amount = Some(input_amount);
+
+	Ok(())
+}
+
+impl<T: Config> Pallet<T> {
+	///Amount is saturated to u128, don't exchange more than u128::MAX at once.
+	pub fn verify_cc_ocw(
+		u_cc: &UnverifiedCollectCoins,
+		collect_coins: &mut Option<CollectCoins<T::Hash, T::Balance>>,
+	) -> VerificationResult<T::Moment> {
+		log::debug!("verifying OCW Collect Coins");
+		let UnverifiedCollectCoins { to, tx_id } = u_cc;
+		let rpc_url = &CONTRACT_CHAIN.rpc_url()?;
+		let tx = rpc::eth_get_transaction(tx_id, rpc_url)?;
+		let tx_receipt = rpc::eth_get_transaction_receipt(tx_id, rpc_url)?;
+		let eth_tip = rpc::eth_get_block_number(rpc_url)?;
+
+		let amount = &mut None;
+		validate_collect_coins(to, amount, &tx_receipt, &tx, eth_tip)?;
```

Comment:
why use a mutable `amount`, couldn't you return `amount` from `validate_collect_coins`?